### PR TITLE
[lld][ELF] Error when deplibs adds new input file after LTO

### LINKED ISSUE #56070: [LTO] Objects brought in by LTO-generated libcalls do not handle deplibs.
We've been seeing failures when attempting to build ICU and FFmpeg on aarch64-...-fuchsia with LTO enabled:
```
ld.lld: error: undefined symbol: _zx_system_get_features
>>> referenced by cpu_model.c:818 (.../cpu_model.c:818)
>>>               cpu_model.c.obj:(init_have_lse_atomics) in archive .../clang/linux-x64/lib/clang/15.0.0/lib/aarch64-unknown-fuchsia/libclang_rt.builtins.a
clang++: error: ld.lld command failed with exit code 1 (use -v to see invocation)
```

We've traced this down to an LTO corner case. LTO code generation can introduce new undefined symbols by emitting external references to libcalls; this can, for example, happen if external atomics are used on aarch64. From the comments in the ELF LTO driver, it appears that it's not generally desirable to pull these in before LTO (except if absolutely necessary, say, if they contain bitcode), since any given libcall may not actually ever be emitted, and this may pull in unnecessary dependencies. Until LTO code generation is finished, the precise set of libcalls needed is unknown, and cannot be determined by examining the bitcode.

The LTO object file seems to be  parsed and added to the link largely if it were an input file; this may pull in other object files to satisfy these new libcall references. This seems to work fine, as far as it goes; the trouble is that the objects files pulled in are only partially processed. In particular, if they contain a `.deplibs` section pointing to another library, then that library is never fully added to the link, which can cause symbols in the object file (`_zx_system_get_features` above) to remain unresolved.

A candidate fix (https://reviews.llvm.org/D127885) is to add `parseFile()` calls to handle any newly-introduced files (such as those produced by .deplibs) as they would have been handled had they been present before LTO occurred.

### PR DESCRIPTION
**Pull request by mysterymath**
Parsing the new input file's symbols might invalidate LTO codegen, but the semantics of deplibs require them to be parsed. Accordingly, report an error unless the file had already been added to the link.

Fixes #56070

### CODE REVIEW

File: lld/test/ELF/deplibs-lto.s [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# REQUIRES: x86
```
Comment (MaskRay) [inline lld/test/ELF/deplibs-lto.s:1]: aarch64

File: lld/test/ELF/deplibs-lto.s [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+# RUN: rm -rf %t && split-file %s %t && cd %t
+# RUN: llvm-mc -filetype=obj -triple=aarch64 -o deplibs.o deplibs.s
+# RUN: llvm-mc -filetype=obj -triple=aarch64 -o foo.o foo.s
+# RUN: llvm-as -o lto.o lto.ll
+# RUN: llvm-ar rc libdeplibs.a deplibs.o
+# RUN: llvm-ar rc libfoo.a foo.o
+
+## LTO emits a libcall (`__aarch64_ldadd4_relax`) that is resolved using a
+## library (libdeplibc.a) that contains a `.deplibs` section pointing to a file
```
Comment (MaskRay) [inline lld/test/ELF/deplibs-lto.s:11]: libdeplibs.a

File: lld/test/ELF/deplibs-lto.s [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+## not yet added to the link.
+
+# RUN: not ld.lld lto.o -u a -L. -ldeplibs 2>&1 | FileCheck %s
+# CHECK: error: input file 'foo.o' added after LTO
+
+## Including the file before LTO prevents the issue.
+
```
Comment (MaskRay) [inline lld/test/ELF/deplibs-lto.s:18]: Delete the blank line. The convention is to have `##` immediately precede associated RUN lines.

File: lld/ELF/Driver.cpp [hunk 4]
```diff
@@ -2980,6 +2981,20 @@ template <class ELFT> void LinkerDriver::link(opt::InputArgList &args) {
   for (const DuplicateSymbol &d : ctx.duplicates)
     reportDuplicate(*d.sym, d.file, d.section, d.value);
 
+  // ELF dependent libraries may have introduced new input files after LTO has
+  // completed. This is an error if the files haven't already been parsed, since
+  // it's no longer legal to change the symbol table by parsing them.
```
Comment (MaskRay) [inline lld/ELF/Driver.cpp:2986]: "no longer legal ..." => changing the symbol table could break the semantic assumption of LTO.

**Review by MaskRay — APPROVED** [review]

File: lld/test/ELF/deplibs-lto.s (hunk 2 above)
Comment (mysterymath) [reply lld/test/ELF/deplibs-lto.s]: Done.

File: lld/test/ELF/deplibs-lto.s (hunk 3 above)
Comment (mysterymath) [reply lld/test/ELF/deplibs-lto.s]: Done.

File: lld/ELF/Driver.cpp (hunk 4 above)
Comment (mysterymath) [reply lld/ELF/Driver.cpp]: Done.

**Review by MaskRay — APPROVED** [review]

**Pull request merged**